File: 0918-maximum-sum-circular-subarray/0918-maximum-sum-circular-subarray.cpp
class Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        int mx=INT_MIN,mn=INT_MAX,sum1=0,sum2=0,Tsum=0;
        for(int i=0;i<nums.size();i++){
            Tsum+=nums[i];
            sum1+=nums[i];
            mx=max(mx,sum1);
            if(sum1<0) sum1=0;
            sum2+=nums[i];
            mn=min(mn,sum2);
            if(sum2>0) sum2=0;
        }
        return mx<0?mx:max(Tsum-mn,mx);
    }
};

File: 0918-maximum-sum-circular-subarray/0918-maximum-sum-circular-subarray.py
class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        if max(nums) <= 0:
            return max(nums)
        max_sum = curr_max = min_sum = curr_min = nums[0] 
        for i in range(1, len(nums)): 
            curr_max = max(nums[i], curr_max + nums[i]) 
            max_sum = max(max_sum, curr_max)
            curr_min = min(nums[i], curr_min + nums[i]) 
            min_sum = min(min_sum, curr_min)
        return max(max_sum, sum(nums) - min_sum)
        